MaxText/layers/gpt3.py

```diff
@@ -0,0 +1,340 @@
+
+      https://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ """
+
+"""Transformer model definition."""
+# pylint: disable=arguments-differ
+# pylint: disable=no-name-in-module
+
+
+from typing import Any, Tuple
+
+from jax.sharding import Mesh
+from jax import lax
+import jax.numpy as jnp
+from jax.ad_checkpoint import checkpoint_name
+
+from flax import linen as nn
+
+from layers import attentions
+from layers import initializers
+from layers import linears
+from layers import models
+
+AttentionOp = attentions.AttentionOp
+
+import common_types
+
+Array = common_types.Array
+Config = common_types.Config
+DType = common_types.DType
+Mesh = common_types.Mesh
+AxisNames = common_types.AxisNames
+BATCH = common_types.BATCH
+LENGTH = common_types.LENGTH
+HEAD = common_types.HEAD
+D_KV = common_types.D_KV
+
+DenseGeneral = linears.DenseGeneral
+NdInitializer = initializers.NdInitializer
+Initializer = initializers.Initializer
+nd_dense_init = initializers.nd_dense_init
+
+
+#-----------------------------------------
+# The Normalization Layer specific for GPT3
+#-----------------------------------------
+
+class Gpt3LayerNorm(nn.Module):
+  """GPT3 Layer normalization operating on the last axis of the input data."""
+  epsilon: float = 1e-6
+  dtype: Any = jnp.float32
+  kernel_axes: Tuple[str, ...] = ()
+  scale_init: Initializer = nn.initializers.zeros
+  use_bias: bool = True
+  reductions_in_fp32: bool = False
+
+  @nn.compact
+  def __call__(self, x: jnp.ndarray) -> jnp.ndarray:
+    """Applies layer normalization on the input."""
+    if self.reductions_in_fp32:
+      x = jnp.asarray(x, jnp.float32)
+    mean = jnp.mean(x, axis=[-1], keepdims=True)
+    var = jnp.mean(jnp.square(x - mean), axis=[-1], keepdims=True)
+    normed_inputs = (x - mean) * lax.rsqrt(var + self.epsilon)
+    if self.reductions_in_fp32:
+      normed_inputs = normed_inputs.astype(self.dtype)
+
+    features = x.shape[-1]
+    scale = self.param(
+        'scale',
+        nn.with_logical_partitioning(self.scale_init, self.kernel_axes),
+        (features,),
+        jnp.float32,
+    )
+
+    scale = jnp.asarray(scale, self.dtype)
+    output = normed_inputs * (scale + 1)
+
+    if self.use_bias:
+      bias = self.param(
+        'bias',
+        nn.with_logical_partitioning(initializers.default_bias_init, self.kernel_axes),
+        (features,),
+        jnp.float32,
+      )
+      bias = jnp.asarray(bias, self.dtype)
+      output += bias
+    return output
+
+
+#-----------------------------------------
+# The Attention Layer specific for GPT3
+#-----------------------------------------
+
+class Gpt3MultiHeadAttention(nn.Module):
```

Comment:
Sorry not understanding why you have your own attention here?

Reply:
> I do hope to make compatible to the original attention class. However, I hope to avoid more complexity to add the following flags for compatibility 
> 1. enable_rope, 2)depth_scaling, 3)use_bias option in layer_norms and projections. 
> In addition to that, there's no differences between input_q, input_kv and projection_query, projection_kv in gpt3 implementation.

Discussed offline. Got accepted.